segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,453 @@
+        retFuture.complete(exists);
+    }
+
+
+    private void syncCreate(String streamSegmentName, Duration timeout, CompletableFuture<SegmentProperties> retVal) {
+        log.info("Creating Segment {}", streamSegmentName);
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            retVal.complete(this.getStreamSegmentInfo(streamSegmentName, timeout).get());
+        } catch (Exception e) {
+            log.info("Exception {} while creating a segment {}", e, streamSegmentName);
+            if (e instanceof FileAlreadyExistsException) {
+                retVal.completeExceptionally(new StreamSegmentExistsException(streamSegmentName, e));
+            } else {
+                retVal.completeExceptionally(e);
+            }
+        }
+    }
+
+    private void syncWrite(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout,
+                           CompletableFuture<Void> retVal) {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (handle.isReadOnly()) {
+            log.info("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            retVal.completeExceptionally(new IllegalArgumentException());
+            return;
+        }
+
+        if (!Files.exists(path)) {
+            retVal.completeExceptionally(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+            return;
+        }
+
+        try {
+            if (!isWritableFile(path)) {
+                retVal.completeExceptionally(new StreamSegmentSealedException(handle.getSegmentName()));
+                return;
+            }
+
+            long fileSize = path.toFile().length();
+            if (fileSize < offset) {
+                retVal.completeExceptionally(new BadOffsetException(handle.getSegmentName(), fileSize, offset));
+            } else {
+                try (FileChannel channel = FileChannel.open(path,
+                        StandardOpenOption.WRITE); ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    channel.force(true);
```

Comment:
Actually the force() here is not necessary as the file handle is closed after each write. This will ensure metadatawrite as well. This is necessary as the assumption is that both data and metadata are transferred reliably to Tier 2 when this function returns.